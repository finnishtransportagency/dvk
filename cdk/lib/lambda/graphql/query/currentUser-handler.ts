import { AppSyncResolverEvent } from 'aws-lambda/trigger/appsync-resolver';
import { CurrentUser } from '../../../../graphql/generated';
import { validateJwtToken } from '../../api/login';
import { log } from '../../logger';

export const handler = async (event: AppSyncResolverEvent<void>): Promise<CurrentUser> => {
  log.info({ headers: event.request.headers }, `currentUser(${event.identity})`);
  const token = event.request.headers['x-iam-accesstoken'];
  const data = event.request.headers['x-iam-data'];
  const jwtToken = await validateJwtToken(token, data || '');
  log.debug('JwtToken: %s', jwtToken);
  return {
    name: jwtToken ? jwtToken['custom:etunimi'] + ' ' + jwtToken['custom:sukunimi'] : '????',
  };
};
