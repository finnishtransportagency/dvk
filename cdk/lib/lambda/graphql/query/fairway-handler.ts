import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { Fairway, QueryFairwayArgs } from '../../../../graphql/generated';
import { log } from '../../logger';
import FairwayDBModel from '../../db/fairwayModel';
import { FairwayService } from '../../api/fairwayService';

const fairwayService = new FairwayService();

export const handler: AppSyncResolverHandler<QueryFairwayArgs, Fairway> = async (event: AppSyncResolverEvent<QueryFairwayArgs>): Promise<Fairway> => {
  log.info(`fairway(${event.arguments.id})`);
  const dbModel = await FairwayDBModel.get(event.arguments.id);
  const apiModel = fairwayService.getFairway(event.arguments.id);
  const fairway: Fairway = {
    id: event.arguments.id,
    name: {
      fi: apiModel.nimiFI,
      sv: apiModel.nimiSV || '',
      en: dbModel?.name || '',
    },
    draft1: apiModel.kulkuSyvyys1,
    draft2: apiModel.kulkuSyvyys2,
    draft3: apiModel.kulkuSyvyys3,
    length: apiModel.pituus,
  };
  return fairway;
};
