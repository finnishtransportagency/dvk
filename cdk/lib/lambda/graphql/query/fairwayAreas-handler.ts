import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda';
import axios from 'axios';
import { Area, Fairway, QueryFairwayCardArgs } from '../../../../graphql/generated';
import { getVatuHeaders, getVatuUrl } from '../../environment';
import { log } from '../../logger';

export type AlueVaylaAPIModel = {
  jnro: number;
  status: number;
  linjaus: number;
  mitoitusNopeus?: number;
  mitoitusNopeus2?: number;
};
export type AlueAPIModel = {
  id: number;
  nimi?: string;
  mitoitusSyvays?: number;
  harausSyvyys?: number;
  vertaustaso?: string;
  n2000MitoitusSyvays?: number;
  n2000HarausSyvyys?: number;
  n2000Vertaustaso?: string;
  suunta?: number;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  omistaja?: string;
  lisatieto?: string;
  vayla?: AlueVaylaAPIModel[];
  tyyppiKoodi?: string;
  tyyppi?: string;
  merkintalajiKoodi?: number;
  merkintalaji?: string;
  liikennointiStatusKoodi?: number;
  liikennointiStatus?: string;
  liikennointiTyyppiKoodi?: number;
  liikenteenTyyppi?: string;
  liikennointiSuuntaKoodi?: number;
  liikennointiSuunta?: string;
  geometria: object;
};

export const handler: AppSyncResolverHandler<QueryFairwayCardArgs, Area[], Fairway> = async (
  event: AppSyncResolverEvent<QueryFairwayCardArgs, Fairway>
): Promise<Area[]> => {
  log.info(`areas(${event.source.id})`);
  const response = await axios.get(`${await getVatuUrl()}/vaylaalueet?jnro=${event.source.id}`, {
    headers: await getVatuHeaders(),
  });
  const areas = response.data as AlueAPIModel[];
  log.debug('areas: %d', areas.length);
  return areas.map((apiArea) => {
    const area: Area = {
      id: apiArea.id,
      name: apiArea.nimi,
      depth: apiArea.mitoitusSyvays,
      draft: apiArea.harausSyvyys,
      referenceLevel: apiArea.vertaustaso,
      n2000depth: apiArea.n2000MitoitusSyvays,
      n2000draft: apiArea.n2000HarausSyvyys,
      n2000ReferenceLevel: apiArea.n2000Vertaustaso,
      direction: apiArea.suunta,
      journalNumber: apiArea.diaariNumero,
      verificationDate: apiArea.vahvistusPaivamaara,
      owner: apiArea.omistaja,
      additionalInformation: apiArea.lisatieto,
      typeCode: apiArea.tyyppiKoodi,
      type: apiArea.tyyppi,
      notationCode: apiArea.merkintalajiKoodi,
      notation: apiArea.merkintalaji,
      operationStatusCode: apiArea.liikennointiStatusKoodi,
      operationStatus: apiArea.liikennointiStatus,
      operationTypeCode: apiArea.liikennointiTyyppiKoodi,
      operationType: apiArea.liikenteenTyyppi,
      operationDirectionCode: apiArea.liikennointiSuuntaKoodi,
      operationDirection: apiArea.liikennointiSuunta,
      geometry: apiArea.geometria,
    };
    area.fairways = apiArea.vayla?.map((apiFairway) => {
      return {
        fairwayId: apiFairway.jnro,
        line: apiFairway.linjaus,
        status: apiFairway.status,
        sizingSpeed: apiFairway.mitoitusNopeus,
        sizingSpeed2: apiFairway.mitoitusNopeus2,
      };
    });
    return area;
  });
};
