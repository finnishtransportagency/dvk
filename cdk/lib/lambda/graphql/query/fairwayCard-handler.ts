import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { FairwayCard, QueryFairwayCardArgs } from '../../../../graphql/generated';
import { log } from '../../logger';
import FairwayCardDBModel from '../../db/fairwayCardDBModel';
import { FairwayService } from '../../api/fairwayService';

const fairwayService = new FairwayService();

export const handler: AppSyncResolverHandler<QueryFairwayCardArgs, FairwayCard | undefined> = async (
  event: AppSyncResolverEvent<QueryFairwayCardArgs>
): Promise<FairwayCard | undefined> => {
  log.info(`fairwayCard(${event.arguments.id})`);
  const dbModel = await FairwayCardDBModel.get(event.arguments.id);
  if (dbModel) {
    const card: FairwayCard = {
      id: dbModel.id,
      name: {
        fi: dbModel.name?.fi || '',
        sv: dbModel.name?.sv || '',
        en: dbModel.name?.en || '',
      },
      fairways: [],
    };
    for (const fairway of dbModel?.fairways || []) {
      card.fairways.push(fairwayService.mapModelsToFairway(undefined, fairway));
    }
    return card;
  }
  return undefined;
};
