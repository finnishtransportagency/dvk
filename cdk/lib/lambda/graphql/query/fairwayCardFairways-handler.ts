import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda';
import { Area, Fairway, FairwayCard, NavigationLine, QueryFairwayCardArgs, RestrictionArea } from '../../../../graphql/generated';
import { log } from '../../logger';
import { AlueAPIModel, fetchVATUByFairwayId, NavigointiLinjaAPIModel, RajoitusAlueAPIModel, VaylaAPIModel } from './vatu';

function mapAPIModelToFairway(apiModel: VaylaAPIModel): Fairway {
  const fairway: Fairway = {
    id: apiModel.jnro,
    name: {
      fi: apiModel.nimiFI || '',
      sv: apiModel.nimiSV || '',
    },
    sizing: {
      additionalInformation: apiModel.lisatieto,
      mareograph: apiModel.mareografi,
      minimumTurningCircle: apiModel.minimiKaantosade,
      minimumWidth: apiModel.minimiLeveys,
      normalTurningCircle: apiModel.normaaliKaantosade,
      normalWidth: apiModel.normaaliLeveys,
      reserveWater: apiModel.varavesi,
    },
    typeCode: apiModel.vaylalajiKoodi,
    type: {
      fi: apiModel.vaylaLajiFI,
      sv: apiModel.vaylaLajiSV,
    },
    area: {
      fi: apiModel.merialueFI,
      sv: apiModel.merialueSV,
    },
    lightingCode: apiModel.valaistusKoodi,
    lighting: {
      fi: apiModel.valaistusFI,
      sv: apiModel.valaistusSV,
    },
    owner: apiModel.omistaja,
    startText: apiModel.alunSeloste,
    endText: apiModel.paatepisteenSeloste,
  };
  fairway.sizingVessels = apiModel.mitoitusalus?.map((vesselModel) => {
    return {
      typeCode: vesselModel.alustyyppiKoodi,
      type: vesselModel.alustyyppi,
      draft: vesselModel.syvays,
      length: vesselModel.pituus,
      width: vesselModel.leveys,
      bodyFactor: vesselModel.runkoTaytelaisyysKerroin,
      size: vesselModel.koko,
    };
  });
  fairway.classifications = apiModel.luokitus?.map((classificationModel) => {
    return {
      type: classificationModel.luokitusTyyppi,
      fairwayClassCode: classificationModel.vaylaluokkaKoodi,
      fairwayClass: {
        fi: classificationModel.vaylaluokkaFI,
        sv: classificationModel.vaylaluokkaSV,
      },
    };
  });
  return fairway;
}

async function getNavigationLineMap(fairwayIds: number[]) {
  const lines = await fetchVATUByFairwayId<NavigointiLinjaAPIModel>(fairwayIds, 'navigointilinjat');
  log.debug('lines: %d', lines.length);
  const lineMap = new Map<number, NavigointiLinjaAPIModel[]>();
  for (const line of lines) {
    for (const lineFairway of line.vayla) {
      if (!lineMap.has(lineFairway.jnro)) {
        lineMap.set(lineFairway.jnro, []);
      }
      lineMap.get(lineFairway.jnro)?.push(line);
    }
  }
  return lineMap;
}

function mapNavigationLines(lines: NavigointiLinjaAPIModel[]) {
  return lines.map((apiLine) => {
    const line: NavigationLine = {
      id: apiLine.id,
      draft: apiLine.mitoitusSyvays,
      depth: apiLine.harausSyvyys,
      referenceLevel: apiLine.vertaustaso,
      n2000draft: apiLine.n2000MitoitusSyvays,
      n2000depth: apiLine.n2000HarausSyvyys,
      n2000ReferenceLevel: apiLine.n2000Vertaustaso,
      direction: apiLine.tosisuunta,
      length: apiLine.pituus,
      additionalInformation: apiLine.lisatieto,
      owner: apiLine.omistaja,
      verificationDate: apiLine.vahvistusPaivamaara,
      journalNumber: apiLine.diaariNumero,
      type: apiLine.tyyppi,
      typeCode: apiLine.tyyppiKoodi,
      geometry: apiLine.geometria,
      fairways: [],
    };
    line.fairways = apiLine.vayla?.map((apiFairway) => {
      return {
        fairwayId: apiFairway.jnro,
        status: apiFairway.status,
        line: apiFairway.linjaus,
      };
    });
    return line;
  });
}

async function getAreaMap(fairwayIds: number[]) {
  const areas = await fetchVATUByFairwayId<AlueAPIModel>(fairwayIds, 'vaylaalueet');
  log.debug('areas: %d', areas.length);
  const areaMap = new Map<number, AlueAPIModel[]>();
  for (const area of areas) {
    for (const areaFairway of area.vayla || []) {
      if (!areaMap.has(areaFairway.jnro)) {
        areaMap.set(areaFairway.jnro, []);
      }
      areaMap.get(areaFairway.jnro)?.push(area);
    }
  }
  return areaMap;
}

function mapAreas(areas: AlueAPIModel[]) {
  return areas.map((apiArea) => {
    const area: Area = {
      id: apiArea.id,
      name: apiArea.nimi,
      draft: apiArea.mitoitusSyvays,
      depth: apiArea.harausSyvyys,
      referenceLevel: apiArea.vertaustaso,
      n2000draft: apiArea.n2000MitoitusSyvays,
      n2000depth: apiArea.n2000HarausSyvyys,
      n2000ReferenceLevel: apiArea.n2000Vertaustaso,
      direction: apiArea.suunta,
      journalNumber: apiArea.diaariNumero,
      verificationDate: apiArea.vahvistusPaivamaara,
      owner: apiArea.omistaja,
      additionalInformation: apiArea.lisatieto,
      typeCode: apiArea.tyyppiKoodi,
      type: apiArea.tyyppi,
      notationCode: apiArea.merkintalajiKoodi,
      notation: apiArea.merkintalaji,
      operationStatusCode: apiArea.liikennointiStatusKoodi,
      operationStatus: apiArea.liikennointiStatus,
      operationTypeCode: apiArea.liikennointiTyyppiKoodi,
      operationType: apiArea.liikenteenTyyppi,
      operationDirectionCode: apiArea.liikennointiSuuntaKoodi,
      operationDirection: apiArea.liikennointiSuunta,
      geometry: apiArea.geometria,
    };
    area.fairways = apiArea.vayla?.map((apiFairway) => {
      return {
        fairwayId: apiFairway.jnro,
        line: apiFairway.linjaus,
        status: apiFairway.status,
        sizingSpeed: apiFairway.mitoitusNopeus,
        sizingSpeed2: apiFairway.mitoitusNopeus2,
        sequenceNumber: apiFairway.vaylaalueenJarjestysNro,
      };
    });
    return area;
  });
}

function mapRestrictionAreas(areas: RajoitusAlueAPIModel[]) {
  return areas.map((apiArea) => {
    const area: RestrictionArea = {
      id: apiArea.id,
      value: apiArea.suuruus,
      presenter: apiArea.esittaja,
      journalNumber: apiArea.diaariNumero,
      verificationDate: apiArea.vahvistusPaivamaara,
      modificationDate: apiArea.muutosPaivamaara,
      startDate: apiArea.alkuPaivamaara,
      endDate: apiArea.loppuPaivamaara,
      status: apiArea.paatosTila,
      source: apiArea.tietolahde,
      location: apiArea.sijainti,
      municipality: apiArea.kunta,
      exception: apiArea.poikkeus,
      geometry: apiArea.geometria,
    };
    area.fairways = apiArea.vayla?.map((apiFairway) => {
      return {
        fairwayId: apiFairway.jnro,
      };
    });
    area.types =
      apiArea.rajoitustyypit?.map((t) => {
        return { code: t.koodi, text: t.rajoitustyyppi };
      }) || [];
    if (apiArea.rajoitustyyppi) {
      area.types.push({ text: apiArea.rajoitustyyppi });
    }
    return area;
  });
}

async function getRestrictionAreaMap(fairwayIds: number[]) {
  const areas = await fetchVATUByFairwayId<RajoitusAlueAPIModel>(fairwayIds, 'rajoitusalueet');
  log.debug('restriction areas: %d', areas.length);
  const areaMap = new Map<number, RajoitusAlueAPIModel[]>();
  for (const area of areas) {
    for (const areaFairway of area.vayla || []) {
      if (!areaMap.has(areaFairway.jnro)) {
        areaMap.set(areaFairway.jnro, []);
      }
      areaMap.get(areaFairway.jnro)?.push(area);
    }
  }
  return areaMap;
}

export const handler: AppSyncResolverHandler<QueryFairwayCardArgs, Fairway[], FairwayCard> = async (
  event: AppSyncResolverEvent<QueryFairwayCardArgs, FairwayCard>
): Promise<Fairway[]> => {
  log.info(`fairwayCardFairways(${event.source.id})`);
  const fairwayMap = new Map<number, Fairway>();
  event.source.fairways.forEach((f) => {
    fairwayMap.set(f.id, f);
  });
  const fairwayIds = event.source.fairways.map((f) => f.id);
  log.debug(`fairwayIds: ${fairwayIds}`);
  const lineMap = await getNavigationLineMap(fairwayIds);
  const areaMap = await getAreaMap(fairwayIds);
  const restrictionAreaMap = await getRestrictionAreaMap(fairwayIds);
  const fairways = await fetchVATUByFairwayId<VaylaAPIModel>(fairwayIds, 'vaylat');
  return fairways.map((apiFairway) => {
    const fairway = fairwayMap.get(apiFairway.jnro);
    log.debug('Fairway: %o', apiFairway);
    return {
      ...mapAPIModelToFairway(apiFairway),
      ...fairway,
      navigationLines: mapNavigationLines(lineMap.get(apiFairway.jnro) || []),
      areas: mapAreas(areaMap.get(apiFairway.jnro) || []),
      restrictionAreas: mapRestrictionAreas(restrictionAreaMap.get(apiFairway.jnro) || []),
    };
  });
};
