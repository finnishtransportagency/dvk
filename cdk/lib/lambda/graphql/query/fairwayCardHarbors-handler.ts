import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda';
import { FairwayCard, Harbor, QueryFairwayCardArgs } from '../../../../graphql/generated';
import { getOptionalCurrentUser } from '../../api/login';
import HarborDBModel from '../../db/harborDBModel';
import { mapHarborDBModelToGraphqlType } from '../../db/modelMapper';
import { log } from '../../logger';

export const handler: AppSyncResolverHandler<QueryFairwayCardArgs, Harbor[], FairwayCard> = async (
  event: AppSyncResolverEvent<QueryFairwayCardArgs, FairwayCard>
): Promise<Harbor[]> => {
  log.info(`fairwayCardHarbors(${event.source.id})`);
  const harborIds = event.source.harbors?.map((h) => h.id);
  log.debug(`harborIds: ${harborIds}`);
  const harbors: Harbor[] = [];
  const user = await getOptionalCurrentUser(event);
  for (const id of harborIds || []) {
    const model = await HarborDBModel.getPublic(id);
    if (model) {
      harbors.push(mapHarborDBModelToGraphqlType(model, user));
    }
  }
  return harbors;
};
