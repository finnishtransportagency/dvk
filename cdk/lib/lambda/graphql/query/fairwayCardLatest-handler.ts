import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { FairwayCard, QueryFairwayCardArgs } from '../../../../graphql/generated';
import { log } from '../../logger';
import FairwayCardDBModel from '../../db/fairwayCardDBModel';
import { getPilotPlaceMap, getPilotRoutes, mapFairwayCardDBModelToGraphqlType } from '../../db/modelMapper';
import { getOptionalCurrentUser } from '../../api/login';
import { FeatureCollection, GeoJsonProperties, Geometry } from 'geojson';

export const handler: AppSyncResolverHandler<QueryFairwayCardArgs, FairwayCard | undefined> = async (
  event: AppSyncResolverEvent<QueryFairwayCardArgs>
): Promise<FairwayCard | undefined> => {
  log.info(`fairwayCard(${event.arguments.id})`);
  const pilotMap = await getPilotPlaceMap();
  const pilotRoutes = (await getPilotRoutes()) as FeatureCollection<Geometry, GeoJsonProperties>;
  const dbModel = await FairwayCardDBModel.getLatest(event.arguments.id);
  if (dbModel) {
    const user = await getOptionalCurrentUser(event);
    return mapFairwayCardDBModelToGraphqlType(dbModel, pilotMap, user, pilotRoutes);
  }
  return undefined;
};