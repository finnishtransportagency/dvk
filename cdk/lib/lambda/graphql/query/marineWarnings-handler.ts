import { GetObjectCommand, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { MarineWarning } from '../../../../graphql/generated';
import { fetchMarineWarnings, parseDateTimes } from '../../api/pooki';
import { getEnvironment } from '../../environment';
import { log } from '../../logger';
import { Readable } from 'stream';

const s3Client = new S3Client({ region: 'eu-west-1' });

function getKey() {
  return 'marinewarning-graphql';
}

function getCacheBucketName() {
  return `featurecache-${getEnvironment()}`;
}

async function cacheResponse(key: string, response: object) {
  const expires = new Date();
  expires.setTime(expires.getTime() + 24 * 60 * 60 * 1000);
  const command = new PutObjectCommand({
    Key: key,
    Bucket: getCacheBucketName(),
    Expires: expires,
    Body: JSON.stringify(response),
  });
  await s3Client.send(command);
  log.debug(`${key} cached`);
}

type CacheResponse = {
  expired: boolean;
  data?: string;
};

async function getFromCache(key: string): Promise<CacheResponse> {
  try {
    const data = await s3Client.send(
      new GetObjectCommand({
        Key: key,
        Bucket: getCacheBucketName(),
      })
    );
    if (data.Body) {
      log.debug(`returning ${key} from cache`);
      const streamToString = (stream: Readable): Promise<string> =>
        new Promise((resolve, reject) => {
          const chunks: Uint8Array[] = [];
          stream.on('data', (chunk: Uint8Array) => chunks.push(chunk));
          stream.on('error', reject);
          stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
        });
      return {
        expired: data.Expires !== undefined && data.Expires.getTime() < Date.now(),
        data: await streamToString(data.Body as Readable),
      };
    }
  } catch (e) {
    // errors ignored also not found
  }
  return { expired: true };
}

export const handler = async (): Promise<MarineWarning[]> => {
  const key = getKey();
  try {
    const resp = await fetchMarineWarnings();
    const warnings = [];
    for (const feature of resp.features) {
      const dates = parseDateTimes(feature);
      const warning: MarineWarning = {
        id: feature.properties?.ID,
        number: feature.properties?.NUMERO || 0,
        area: { fi: feature.properties?.ALUEET_FI, sv: feature.properties?.ALUEET_SV, en: feature.properties?.ALUEET_EN },
        type: { fi: feature.properties?.TYYPPI_FI, sv: feature.properties?.TYYPPI_SV, en: feature.properties?.TYYPPI_EN },
        location: { fi: feature.properties?.SIJAINTI_FI, sv: feature.properties?.SIJAINTI_SV, en: feature.properties?.SIJAINTI_EN },
        description: { fi: feature.properties?.SISALTO_FI, sv: feature.properties?.SISALTO_SV, en: feature.properties?.SISALTO_EN },
        startDateTime: dates.startDateTime,
        endDateTime: dates.endDateTime,
        dateTime: dates.dateTime || 0,
        notifier: feature.properties?.TIEDOKSIANTAJA,
        equipmentId: Number(feature.properties?.TURVALAITE_TXT?.match(/\d.*/)[0]),
        lineId: Number(feature.properties?.NAVIGOINTILINJA_TXT?.match(/\d.*/)[0]),
        areaId: Number(feature.properties?.VAYLAALUE_TXT?.match(/\d.*/)[0]),
      };
      warnings.push(warning);
    }
    await cacheResponse(key, warnings);
    return warnings;
  } catch (e) {
    log.error('Getting marine warnings failed: %s', e);
    const cacheResponseData = await getFromCache(key);
    if (cacheResponseData.data) {
      log.warn('Returning expired response from s3 cache');
      return JSON.parse(cacheResponseData.data);
    } else {
      throw e;
    }
  }
};
