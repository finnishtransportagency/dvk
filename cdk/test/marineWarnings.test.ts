import { mockClient } from 'aws-sdk-client-mock';
import { handler } from '../lib/lambda/graphql/query/marineWarnings-handler';
import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { sdkStreamMixin } from '@smithy/util-stream';
import { FeatureCollection } from 'geojson';
import { createReadStream } from 'fs';

const s3Mock = mockClient(S3Client);

jest.mock('../lib/lambda/environment', () => ({
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getHeaders: () => {},
}));

const warnings: FeatureCollection = {
  type: 'FeatureCollection',
  features: [
    {
      type: 'Feature',
      properties: {
        ID: 24822,
        ALUEET_FI: 'PERÄMERI',
        ALUEET_SV: 'BOTTENVIKEN',
        ALUEET_EN: 'BAY OF BOTHNIA',
        NUMERO: 136,
        SIJAINTI_FI: 'OULU - KEMI VÄYLÄ',
        SIJAINTI_SV: 'ULEÅBORG - KEMI FARLED',
        SIJAINTI_EN: 'OULU - KEMI FAIRWAY',
        SISALTO_FI: '\r\nREUNAMERKKI KRIISINKIVI NR 11133\r\nPAIKASSA  65-12.69N 025-04.00E\r\nEI TOIMINNASSA KORJAUSTÖIDEN VUOKSI',
        SISALTO_SV: '\r\nRANDMÄRKE KRIISINKIVI NR 11133\r\nI POS 65-12.69N 025-04.00E\r\nUR FUNKTION PÅ GRUND AV UNDERHÅLLSARBETE',
        SISALTO_EN: '\r\nEDGE MARK KRIISINKIVI NR 11133\r\nIN POS 65-12.69N 025-04.00E\r\nOUT OF ORDER DUE TO MAITENANCE WORK',
        PAIVAYS: '12.7.2023 12:00',
        TYYPPI_FI: 'LOCAL WARNING',
        TYYPPI_SV: 'LOCAL WARNING',
        TYYPPI_EN: 'LOCAL WARNING',
        VOIMASSA_ALKAA: '2023-07-12 12:00:00',
        VOIMASSA_PAATTYY: null,
        VALITTUKOHDE_TOOLTIP: 'TL:11133 Kriisinkivi, Oikeareunamerkki [Vahvistettu]',
        VIRTUAALINENTURVALAITE: 0,
        NAVTEX: 0,
        TALLENNUSPAIVA: '2023-07-12 11:48:32',
        TURVALAITE_TXT:
          'TLNUMERO:11133\r\nALALAJI:Kiinteä\r\nLAJI:Oikea\r\nTYYPPI:Reunamerkki\r\nNIMIS:Kriisinkivi\r\nVAYLAN_NIMI:Oulu - Kemi väylä\r\nTLNUMERO:11133\r\nALALAJI:Kiinteä\r\nLAJI:Oikea\r\nTYYPPI:Reunamerkki\r\nNIMIS:Kriisinkivi\r\nVAYLAN_NIMI:Oulu - Kemi väylä\r\n',
        VAYLAALUE_TXT: null,
        NAVIGOINTILINJA_TXT: null,
        ANTOPAIVA: '12.07.2023',
        TIEDOKSIANTAJA: 'FINTRAFFIC/BVTS/OFK/KD',
      },
      geometry: {
        type: 'Point',
        coordinates: [2790416.9999999967, 9625463.0051028635],
      },
    },
    {
      type: 'Feature',
      properties: {
        ID: 24803,
        ALUEET_FI: 'PERÄMERI',
        ALUEET_SV: 'BOTTENVIKEN',
        ALUEET_EN: 'BAY OF BOTHNIA',
        NUMERO: 135,
        SIJAINTI_FI: 'KEMI AJOS VÄYLÄ',
        SIJAINTI_SV: 'KEMI AJOS FARLED',
        SIJAINTI_EN: 'KEMI AJOS FAIRWAY',
        SISALTO_FI:
          '\r\nKEMIN AJOKSEN VÄYLÄLLÄ SUORITETAAN KELLUVIEN TURVALAITTEIDEN MUUTOSTÖITÄ 10.7-30.7.2023\r\n\r\nDATE AND TIME\r\n080930 UTC OF JUL',
        SISALTO_SV:
          '\r\nFÖRÄNDRINGSARBETE AV FLYTANDE NAVIGATIONS ANORDNINGAR UTFÖRS PÅ KEMI AJOS FARLED 10.7-30.7.2023\r\n\r\nDATE AND TIME\r\n080930 UTC OF JUL',
        SISALTO_EN:
          '\r\nALTERATION OF FLOATING NAVIGATIONAL AIDS ON KEMI AJOS FAIRWAY WILL BE CONDUCTED BETWEEN 10.7-30.7.2023\r\n\r\nDATE AND TIME\r\n080930 UTC OF JUL',
        PAIVAYS: '8.7.2023 12:30',
        TYYPPI_FI: 'LOCAL WARNING',
        TYYPPI_SV: 'LOCAL WARNING',
        TYYPPI_EN: 'LOCAL WARNING',
        VOIMASSA_ALKAA: '2023-07-10',
        VOIMASSA_PAATTYY: null,
        VALITTUKOHDE_TOOLTIP: null,
        VIRTUAALINENTURVALAITE: 0,
        NAVTEX: 0,
        TALLENNUSPAIVA: '2023-07-08 12:51:30',
        TURVALAITE_TXT: null,
        VAYLAALUE_TXT: null,
        NAVIGOINTILINJA_TXT: null,
        ANTOPAIVA: '08.07.2023',
        TIEDOKSIANTAJA: 'LEHTOLA/FINTRAFFIC/OFK/MMH',
      },
      geometry: {
        type: 'LineString',
        coordinates: [
          [2665506.9999999958, 9658753.0050714873],
          [2678017.0000000014, 9671869.0050591156],
          [2696213.0000000028, 9685819.0050459672],
          [2720701.0000000047, 9716222.0050173067],
          [2728618.9999999949, 9745420.00498979],
        ],
      },
    },
  ],
};

let throwError = false;
jest.mock('../lib/lambda/api/axios', () => ({
  fetchMarineWarnings: () => {
    if (throwError) {
      throw new Error('Fetching from Pooki api failed');
    }
    return warnings;
  },
}));

beforeEach(() => {
  jest.resetAllMocks();
  s3Mock.reset();
  throwError = false;
});

it('should get warnings from api', async () => {
  const stream = sdkStreamMixin(createReadStream('./test/data/warningsgraphql.json'));
  const expires = new Date();
  expires.setTime(expires.getTime() + 1 * 60 * 60 * 1000);
  s3Mock.on(GetObjectCommand).resolves({ Body: stream, ExpiresString: expires.toString() });
  const response = await handler();
  expect(response.length).toBe(2);
  expect(response).toMatchSnapshot();
});

it('should get warnings from cache when api call fails', async () => {
  const stream = sdkStreamMixin(createReadStream('./test/data/warningsgraphql.json'));
  const expires = new Date();
  expires.setTime(expires.getTime() - 1 * 60 * 60 * 1000);
  s3Mock.on(GetObjectCommand).resolves({ Body: stream, ExpiresString: expires.toString() });
  throwError = true;
  const response = await handler();
  expect(response.length).toBe(1);
  expect(response).toMatchSnapshot();
});
