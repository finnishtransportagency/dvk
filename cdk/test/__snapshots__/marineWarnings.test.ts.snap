// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get warnings from api 1`] = `
[
  {
    "area": {
      "en": "BAY OF BOTHNIA",
      "fi": "PERÄMERI",
      "sv": "BOTTENVIKEN",
    },
    "areaId": NaN,
    "dateTime": 1689152400000,
    "description": {
      "en": "
EDGE MARK KRIISINKIVI NR 11133
IN POS 65-12.69N 025-04.00E
OUT OF ORDER DUE TO MAITENANCE WORK",
      "fi": "
REUNAMERKKI KRIISINKIVI NR 11133
PAIKASSA  65-12.69N 025-04.00E
EI TOIMINNASSA KORJAUSTÖIDEN VUOKSI",
      "sv": "
RANDMÄRKE KRIISINKIVI NR 11133
I POS 65-12.69N 025-04.00E
UR FUNKTION PÅ GRUND AV UNDERHÅLLSARBETE",
    },
    "endDateTime": NaN,
    "equipmentId": 11133,
    "id": 24822,
    "lineId": NaN,
    "location": {
      "en": "OULU - KEMI FAIRWAY",
      "fi": "OULU - KEMI VÄYLÄ",
      "sv": "ULEÅBORG - KEMI FARLED",
    },
    "notifier": "FINTRAFFIC/BVTS/OFK/KD",
    "number": 136,
    "startDateTime": 1689152400000,
    "type": {
      "en": "LOCAL WARNING",
      "fi": "LOCAL WARNING",
      "sv": "LOCAL WARNING",
    },
  },
  {
    "area": {
      "en": "BAY OF BOTHNIA",
      "fi": "PERÄMERI",
      "sv": "BOTTENVIKEN",
    },
    "areaId": NaN,
    "dateTime": 1688808600000,
    "description": {
      "en": "
ALTERATION OF FLOATING NAVIGATIONAL AIDS ON KEMI AJOS FAIRWAY WILL BE CONDUCTED BETWEEN 10.7-30.7.2023

DATE AND TIME
080930 UTC OF JUL",
      "fi": "
KEMIN AJOKSEN VÄYLÄLLÄ SUORITETAAN KELLUVIEN TURVALAITTEIDEN MUUTOSTÖITÄ 10.7-30.7.2023

DATE AND TIME
080930 UTC OF JUL",
      "sv": "
FÖRÄNDRINGSARBETE AV FLYTANDE NAVIGATIONS ANORDNINGAR UTFÖRS PÅ KEMI AJOS FARLED 10.7-30.7.2023

DATE AND TIME
080930 UTC OF JUL",
    },
    "endDateTime": NaN,
    "equipmentId": NaN,
    "id": 24803,
    "lineId": NaN,
    "location": {
      "en": "KEMI AJOS FAIRWAY",
      "fi": "KEMI AJOS VÄYLÄ",
      "sv": "KEMI AJOS FARLED",
    },
    "notifier": "LEHTOLA/FINTRAFFIC/OFK/MMH",
    "number": 135,
    "startDateTime": 1688936400000,
    "type": {
      "en": "LOCAL WARNING",
      "fi": "LOCAL WARNING",
      "sv": "LOCAL WARNING",
    },
  },
]
`;

exports[`should get warnings from cache when api call fails 1`] = `
[
  {
    "area": {
      "en": "ARCHIPELAGO SEA",
      "fi": "SAARISTOMERI",
      "sv": "SKÄRGÅRDSHAVET",
    },
    "areaId": 254254,
    "dateTime": 1682629200000,
    "description": {
      "en": null,
      "fi": "DVK testaa",
      "sv": null,
    },
    "endDateTime": 1685480400000,
    "equipmentId": null,
    "id": 24303,
    "lineId": null,
    "location": {
      "en": "Hessund-Parainen veneväylä 3545, syvyys: 1,5",
      "fi": "Hessund-Parainen veneväylä 3545, syvyys: 1,5",
      "sv": "Hessund-Parainen veneväylä 3545, syvyys: 1,5",
    },
    "notifier": "DVK",
    "number": 7777,
    "startDateTime": 1682629200000,
    "type": {
      "en": "LOCAL WARNING",
      "fi": "LOCAL WARNING",
      "sv": "LOCAL WARNING",
    },
  },
]
`;
